llvm/test/CodeGen/AMDGPU/shl64_reduce.ll

```diff
@@ -0,0 +1,1318 @@
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:28
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[6:7], v31, v[6:7]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:36
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[8:9], v31, v[8:9]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:44
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[10:11], v31, v[10:11]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:52
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[12:13], v31, v[12:13]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:60
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[14:15], v31, v[14:15]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:68
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[16:17], v31, v[16:17]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:76
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[18:19], v31, v[18:19]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:84
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[20:21], v31, v[20:21]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:92
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[22:23], v31, v[22:23]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:100
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[24:25], v31, v[24:25]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:108
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[26:27], v31, v[26:27]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32 offset:116
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v31, 16, v31
+; CHECK-NEXT:    v_lshlrev_b64 v[28:29], v31, v[28:29]
+; CHECK-NEXT:    buffer_load_dword v31, off, s[0:3], s32
+; CHECK-NEXT:    buffer_load_dword v32, off, s[0:3], s32 offset:124
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    v_or_b32_e32 v32, 16, v32
+; CHECK-NEXT:    v_lshlrev_b64 v[30:31], v32, v[30:31]
+; CHECK-NEXT:    s_setpc_b64 s[30:31]
+  %or = or <16 x i64> %shift_amt, splat (i64 16)
+  %shl = shl <16 x i64> %arg0, %or
+  ret <16 x i64> %shl
+}
+
+; test inreg
+
+define i64 @shl_or16_inreg(i64 noundef %arg0, i64 inreg %shift_amt) {
```

Comment:
I usually add s and v to indicate the operand types. Inreg doesn't tell me much. Also, the most interesting case requires testing both inputs are inreg so the whole computation is scalar. This is also using a vector return type

Reply:
Used inreg for both inputs.  Changed function name to use "sgpr" suffix.  Could not find a way to return result in SGPR.